pyomo/contrib/gdpopt/solve_subproblem.py

```diff
@@ -0,0 +1,329 @@
+        'fixed binaries and logical realizations.')
+
+    results = configure_and_call_solver(nlp_model, config.nlp_solver,
+                                        config.nlp_solver_args, 'NLP', timing,
+                                        config.time_limit)
+
+    return process_nonlinear_problem_results(results, nlp_model, 'NLP', config)
+
+def solve_MINLP(util_block, config, timing):
+    """Solve the MINLP subproblem."""
+    config.logger.debug(
+        "Solving MINLP subproblem for fixed logical realizations."
+    )
+    model = util_block.model()
+
+    minlp_solver = SolverFactory(config.minlp_solver)
+    if not minlp_solver.available():
+        raise RuntimeError("MINLP solver %s is not available." %
+                           config.minlp_solver)
+
+    results = configure_and_call_solver(model, config.minlp_solver,
+                                        config.minlp_solver_args, 'MINLP',
+                                        timing, config.time_limit)
+    subprob_termination = process_nonlinear_problem_results(results, model,
+                                                            'MINLP', config)
+
+    return subprob_termination
+
+def detect_unfixed_discrete_vars(model):
+    """Detect unfixed discrete variables in use on the model."""
+    var_set = ComponentSet()
+    for constr in model.component_data_objects(
+            Constraint, active=True, descend_into=True):
+        var_set.update(
+            v for v in EXPR.identify_variables(
+                constr.body, include_fixed=False)
+            if not v.is_continuous())
+    for obj in model.component_data_objects(Objective, active=True):
+        var_set.update(v for v in EXPR.identify_variables(obj.expr,
+                                                          include_fixed=False)
+                       if not v.is_continuous())
+    return var_set
+
+@contextmanager
+def preprocess_subproblem(util_block, config):
+    """Applies preprocessing transformations to the model."""
+    m = util_block.model()
+
+    # Save bounds so we can restore them
+    original_bounds = ComponentMap()
+    unfixed_vars = []
+    for cons in m.component_data_objects(Constraint, active=True,
+                                         descend_into=Block):
+        for v in EXPR.identify_variables(cons.expr):
+            if v not in original_bounds.keys():
+                original_bounds[v] = (v.lb, v.ub)
+                if not v.fixed:
+                    unfixed_vars.append(v)
+    # We could miss if there is a variable that only appears in the
+    # objective, but its bounds are not going to get changed anyway if
+    # that's the case.
+    constraints_modified = {}
+    constraints_deactivated = []
+    not_infeas = True
+
+    try:
+        # First do FBBT
+        fbbt(m, integer_tol=config.integer_tolerance,
+             feasibility_tol=config.constraint_tolerance,
+             max_iter=config.max_fbbt_iterations)
```

Comment:
FBBT can be quite slow for some problems (especially if `max_fbbt_iterations` is above 2 or 3). I would try use the implementation in Appsi if it is available. I probably need to add an `available` method first though.